File: services/client/src/forms/ir/swaption.js
import { sensitivityGroup } from '../shared/sensitivity-fields'

export const swaptionForm = {
  name: 'Swaption',
  category: 'IR',
  api: 'calculateSwaption',
  groups: [
    {
      title: 'Option Details',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'evaluation_date',
          type: 'date',
          label: 'Evaluation Date',
          required: true,
          defaultValue: new Date('2016-02-05'),
          helpText: 'The date at which the swaption will be valued'
        },
        {
          key: 'exercise_style',
          type: 'dropdown',
          label: 'Exercise Style',
          required: true,
          defaultValue: 'European',
          options: [
            { label: 'European', value: 'European' },
            { label: 'Bermudan', value: 'Bermudan' },
            { label: 'American', value: 'American' }
          ],
          helpText:
            'European: single exercise date; Bermudan: multiple discrete dates; American: any date in a period'
        },
        {
          key: 'exercise_dates',
          type: 'string',
          label: 'Exercise Date(s)',
          required: true,
          defaultValue: '2036-02-20',
          helpText:
            'Single date for European (YYYY-MM-DD), multiple comma-separated dates for Bermudan/American (e.g., 2033-02-20,2034-02-20)'
        },
        {
          key: 'long_short',
          type: 'dropdown',
          label: 'Position',
          required: true,
          defaultValue: 'Long',
          options: [
            { label: 'Long (Buyer)', value: 'Long' },
            { label: 'Short (Seller)', value: 'Short' }
          ],
          helpText:
            'Long: buy the swaption (right to enter swap); Short: sell the swaption (obligation to enter swap)'
        },
        {
          key: 'settlement_type',
          type: 'dropdown',
          label: 'Settlement',
          required: true,
          defaultValue: 'Cash',
          options: [
            { label: 'Cash', value: 'Cash' },
            { label: 'Physical', value: 'Physical' }
          ],
          helpText:
            'Cash: cash-settled based on swap value; Physical: actual delivery of underlying swap'
        },
        {
          key: 'payer_receiver',
          type: 'dropdown',
          label: 'Swaption Type',
          required: true,
          defaultValue: 'Payer',
          options: [
            { label: 'Payer (Pay Fixed)', value: 'Payer' },
            { label: 'Receiver (Receive Fixed)', value: 'Receiver' }
          ],
          helpText:
            'Payer: option to pay fixed and receive floating; Receiver: option to receive fixed and pay floating'
        }
      ]
    },
    {
      title: 'Premium (Optional)',
      collapsible: true,
      initiallyCollapsed: true,
      fields: [
        {
          key: 'premium_amount',
          type: 'float',
          label: 'Premium Amount',
          required: false,
          step: 1000,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          helpText: 'The upfront premium paid for the swaption (optional)',
          validators: [
            (value) => {
              if (value !== null && value !== undefined && value < 0) {
                return 'Premium amount cannot be negative'
              }
              return true
            }
          ]
        },
        {
          key: 'premium_currency',
          type: 'dropdown',
          label: 'Premium Currency',
          required: false,
          options: [
            { label: 'USD - US Dollar', value: 'USD' },
            { label: 'EUR - Euro', value: 'EUR' },
            { label: 'GBP - British Pound', value: 'GBP' },
            { label: 'JPY - Japanese Yen', value: 'JPY' },
            { label: 'AUD - Australian Dollar', value: 'AUD' },
            { label: 'CAD - Canadian Dollar', value: 'CAD' },
            { label: 'CHF - Swiss Franc', value: 'CHF' }
          ],
          filter: true,
          helpText:
            'Currency of the premium (required if premium amount is specified)'
        },
        {
          key: 'premium_pay_date',
          type: 'date',
          label: 'Premium Pay Date',
          required: false,
          helpText:
            'Date when the premium is paid (required if premium amount is specified)'
        }
      ]
    },
    {
      title: 'Underlying Swap - General',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'currency',
          type: 'dropdown',
          label: 'Currency',
          required: true,
          defaultValue: 'EUR',
          options: [
            { label: 'USD - US Dollar', value: 'USD' },
            { label: 'EUR - Euro', value: 'EUR' },
            { label: 'GBP - British Pound', value: 'GBP' },
            { label: 'JPY - Japanese Yen', value: 'JPY' },
            { label: 'AUD - Australian Dollar', value: 'AUD' },
            { label: 'CAD - Canadian Dollar', value: 'CAD' },
            { label: 'CHF - Swiss Franc', value: 'CHF' }
          ],
          filter: true,
          helpText: 'The currency in which the underlying swap is denominated'
        },
        {
          key: 'notional',
          type: 'float',
          label: 'Notional Amount',
          required: true,
          defaultValue: 10000000,
          step: 1000,
          minFractionDigits: 2,
          maxFractionDigits: 2,
          helpText: 'The notional amount of the underlying swap',
          validators: [
            (value) => {
              if (value <= 0) {
                return 'Notional amount must be greater than 0'
              }
              return true
            }
          ]
        },
        {
          key: 'start_date',
          type: 'date',
          label: 'Swap Start Date',
          required: true,
          defaultValue: new Date('2033-02-20'),
          helpText:
            'The effective date when the underlying swap begins (must be after exercise date for European)',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.evaluation_date) return true
              if (value < formData.evaluation_date) {
                return 'Swap start date cannot be before evaluation date'
              }
              return true
            }
          ]
        },
        {
          key: 'end_date',
          type: 'date',
          label: 'Swap Maturity Date',
          required: true,
          defaultValue: new Date('2043-02-21'),
          helpText: 'The date when the underlying swap matures',
          validators: [
            (value, formData) => {
              if (!value || !formData || !formData.start_date) return true
              if (value <= formData.start_date) {
                return 'Swap maturity date must be after swap start date'
              }
              return true
            }
          ]
        },
        {
          key: 'calendar',
          type: 'dropdown',
          label: 'Business Day Calendar',
          required: true,
          defaultValue: 'TARGET',
          options: [
            { label: 'US - United States', value: 'US' },
            { label: 'TARGET - European Central Bank', value: 'TARGET' },
            { label: 'UK - United Kingdom', value: 'UK' },
            { label: 'JP - Japan', value: 'JP' },
            { label: 'AU - Australia', value: 'AU' },
            { label: 'CA - Canada', value: 'CA' },
            { label: 'CH - Switzerland', value: 'CH' }
          ],
          filter: true,
          helpText: 'The business day calendar used for date adjustments'
        },
        {
          key: 'payment_convention',
          type: 'dropdown',
          label: 'Payment Convention',
          required: true,
          defaultValue: 'MF',
          options: [
            { label: 'Following', value: 'Following' },
            { label: 'Modified Following', value: 'MF' },
            { label: 'Preceding', value: 'Preceding' },
            { label: 'Modified Preceding', value: 'ModifiedPreceding' }
          ],
          filter: true,
          helpText: 'Business day convention for payment dates'
        }
      ]
    },
    {
      title: 'Underlying Swap - Fixed Leg',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'fixed_rate',
          type: 'float',
          label: 'Fixed Rate',
          required: true,
          defaultValue: 0.02,
          min: 0,
          max: 1,
          step: 0.0001,
          minFractionDigits: 4,
          maxFractionDigits: 6,
          helpText: 'The fixed rate in decimal form (e.g., 0.02 for 2%)',
          validators: [
            (value) => {
              if (value < 0) {
                return 'Fixed rate cannot be negative'
              }
              if (value > 1) {
                return 'Fixed rate cannot exceed 1.0 (100%)'
              }
              return true
            }
          ]
        },
        {
          key: 'fixed_tenor',
          type: 'dropdown',
          label: 'Fixed Leg Payment Frequency',
          required: true,
          defaultValue: '1Y',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently fixed rate payments are made'
        },
        {
          key: 'fixed_day_counter',
          type: 'dropdown',
          label: 'Fixed Leg Day Count Convention',
          required: true,
          defaultValue: 'ACT/ACT',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText:
            'The convention used to calculate the fraction of a year for fixed leg payments'
        }
      ]
    },
    {
      title: 'Underlying Swap - Floating Leg',
      collapsible: true,
      initiallyCollapsed: false,
      fields: [
        {
          key: 'floating_index',
          type: 'dropdown',
          label: 'Floating Rate Index',
          required: true,
          defaultValue: 'EUR-EURIBOR-3M',
          options: [
            { label: 'USD-LIBOR-3M', value: 'USD-LIBOR-3M' },
            { label: 'USD-LIBOR-6M', value: 'USD-LIBOR-6M' },
            { label: 'USD-SOFR', value: 'USD-SOFR' },
            { label: 'EUR-EURIBOR-3M', value: 'EUR-EURIBOR-3M' },
            { label: 'EUR-EURIBOR-6M', value: 'EUR-EURIBOR-6M' },
            { label: 'EUR-ESTR', value: 'EUR-ESTR' },
            { label: 'GBP-LIBOR-3M', value: 'GBP-LIBOR-3M' },
            { label: 'GBP-LIBOR-6M', value: 'GBP-LIBOR-6M' },
            { label: 'GBP-SONIA', value: 'GBP-SONIA' },
            { label: 'JPY-LIBOR-3M', value: 'JPY-LIBOR-3M' },
            { label: 'JPY-LIBOR-6M', value: 'JPY-LIBOR-6M' },
            { label: 'JPY-TONAR', value: 'JPY-TONAR' },
            { label: 'AUD-BBSW-3M', value: 'AUD-BBSW-3M' },
            { label: 'AUD-BBSW-6M', value: 'AUD-BBSW-6M' },
            { label: 'CAD-CDOR-3M', value: 'CAD-CDOR-3M' },
            { label: 'CHF-LIBOR-3M', value: 'CHF-LIBOR-3M' },
            { label: 'CHF-LIBOR-6M', value: 'CHF-LIBOR-6M' }
          ],
          filter: true,
          helpText: 'The floating rate index used for the floating leg'
        },
        {
          key: 'floating_tenor',
          type: 'dropdown',
          label: 'Floating Leg Payment Frequency',
          required: true,
          defaultValue: '3M',
          options: [
            { label: 'Annual', value: '1Y' },
            { label: 'Semi-Annual', value: '6M' },
            { label: 'Quarterly', value: '3M' },
            { label: 'Monthly', value: '1M' }
          ],
          filter: true,
          helpText: 'How frequently floating rate payments are made'
        },
        {
          key: 'floating_day_counter',
          type: 'dropdown',
          label: 'Floating Leg Day Count Convention',
          required: true,
          defaultValue: 'A360',
          options: [
            { label: 'Actual/360', value: 'A360' },
            { label: 'Actual/365', value: 'A365' },
            { label: 'Actual/Actual', value: 'ACT/ACT' },
            { label: '30/360', value: '30/360' }
          ],
          filter: true,
          helpText:
            'The convention used to calculate the fraction of a year for floating leg payments'
        },
        {
          key: 'floating_spread',
          type: 'float',
          label: 'Floating Spread',
          required: true,
          defaultValue: 0.0,
          step: 0.0001,
          minFractionDigits: 4,
          maxFractionDigits: 6,
          helpText:
            'Spread over the floating index in decimal form (e.g., 0.001 for 10bp)'
        },
        {
          key: 'is_in_arrears',
          type: 'boolean',
          label: 'Is In Arrears',
          required: false,
          defaultValue: false,
          helpText: 'Whether the fixing is in arrears (fixed at end of period)'
        },
        {
          key: 'fixing_days',
          type: 'number',
          label: 'Fixing Days',
          required: true,
          defaultValue: 2,
          min: 0,
          max: 10,
          helpText:
            'Number of business days between rate fixing and payment period start',
          validators: [
            (value) => {
              if (value < 0) {
                return 'Fixing days cannot be negative'
              }
              if (value > 10) {
                return 'Fixing days cannot exceed 10'
              }
              return true
            }
          ]
        }
      ]
    },
    sensitivityGroup
  ]
}

File: services/server/project/qld/ORETrade/ir/swaption.py
"""
Swaption Trade Implementation using ORE.

This module provides Interest Rate Swaption pricing using the Open Risk Engine (ORE).
Supports European, Bermudan, and American exercise styles with Cash and Physical settlement.
"""

import pandas as pd
import math
import project.qld.engine.parser.date as dateParser

from ..ore_base_trade import OREBaseTrade


class Swaption(OREBaseTrade):
    """
    Interest Rate Swaption trade priced using ORE.

    A swaption is an option to enter into an interest rate swap at a future date.
    The underlying swap exchanges fixed rate payments for floating rate payments.
    """

    __name__ = "Swaption"

    def __init__(self, evaluation_date,
                 # Option parameters
                 exercise_style, exercise_dates, long_short, settlement_type,
                 payer_receiver,
                 # Premium parameters (optional)
                 premium_amount=None, premium_currency=None, premium_pay_date=None,
                 # Underlying swap parameters
                 currency=None, notional=None, start_date=None, end_date=None,
                 # Fixed leg parameters
                 fixed_rate=None, fixed_tenor=None, fixed_day_counter=None,
                 # Floating leg parameters
                 floating_index=None, floating_tenor=None, floating_day_counter=None,
                 floating_spread=0.0,
                 # Common parameters
                 calendar='TARGET', payment_convention='MF',
                 is_in_arrears='false', fixing_days=2):
        """
        Initialize Swaption trade.

        Args:
            evaluation_date: Valuation date

            Option parameters:
            exercise_style: 'European', 'Bermudan', or 'American'
            exercise_dates: Single date for European, multiple dates for Bermudan/American
                           Can be a string (single date) or comma-separated string (multiple dates)
            long_short: 'Long' (buyer) or 'Short' (seller)
            settlement_type: 'Cash' or 'Physical'
            payer_receiver: 'Payer' (pay fixed) or 'Receiver' (receive fixed)

            Premium (optional):
            premium_amount: Premium amount paid/received
            premium_currency: Premium currency
            premium_pay_date: Premium payment date

            Underlying swap:
            currency: Swap currency (e.g., 'EUR', 'USD')
            notional: Notional amount
            start_date: Swap start date (exercise date for European)
            end_date: Swap end date (maturity)

            Fixed leg:
            fixed_rate: Fixed rate (in decimal, e.g., 0.02 for 2%)
            fixed_tenor: Payment frequency (e.g., '1Y', '6M')
            fixed_day_counter: Day count convention (e.g., 'ACT/ACT', 'A360')

            Floating leg:
            floating_index: Floating rate index (e.g., 'EUR-EURIBOR-6M')
            floating_tenor: Payment frequency (e.g., '6M', '3M')
            floating_day_counter: Day count convention
            floating_spread: Spread over index (in decimal, default 0.0)

            Common:
            calendar: Business day calendar (default 'TARGET')
            payment_convention: Payment convention (default 'MF')
            is_in_arrears: Whether fixing is in arrears (default 'false')
            fixing_days: Number of fixing days (default 2)
        """
        super().__init__(evaluation_date)

        # Option parameters
        self.exercise_style = exercise_style
        self.exercise_dates = exercise_dates
        self.long_short = long_short
        self.settlement_type = settlement_type
        self.payer_receiver = payer_receiver

        # Premium parameters
        self.premium_amount = premium_amount
        self.premium_currency = premium_currency
        self.premium_pay_date = premium_pay_date

        # Underlying swap parameters
        self.currency = currency
        self.notional = notional
        self.start_date = start_date
        self.end_date = end_date

        # Fixed leg
        self.fixed_rate = fixed_rate
        self.fixed_tenor = fixed_tenor
        self.fixed_day_counter = fixed_day_counter

        # Floating leg
        self.floating_index = floating_index
        self.floating_tenor = floating_tenor
        self.floating_day_counter = floating_day_counter
        self.floating_spread = floating_spread

        # Common
        self.calendar = calendar
        self.payment_convention = payment_convention
        self.is_in_arrears = is_in_arrears
        self.fixing_days = fixing_days

        self._parse_input()
        self._validate_input()

    def _parse_input(self):
        """Parse and convert string inputs to appropriate types."""
        self.evaluation_date = dateParser.parse_date(self.evaluation_date)

        # Parse option parameters
        self.exercise_style = str(self.exercise_style).capitalize()
        self.long_short = str(self.long_short).capitalize()
        self.settlement_type = str(self.settlement_type).capitalize()
        self.payer_receiver = str(self.payer_receiver).capitalize()

        # Parse exercise dates - can be single date or comma-separated dates
        if isinstance(self.exercise_dates, str):
            # Already a string, will be handled in validation
            pass
        elif isinstance(self.exercise_dates, list):
            # Convert list to comma-separated string
            self.exercise_dates = ','.join([str(d) for d in self.exercise_dates])

        # Parse underlying swap parameters
        if self.currency:
            self.currency = self.currency.upper()
        if self.notional:
            self.notional = float(self.notional)
        if self.start_date:
            self.start_date = dateParser.parse_date(self.start_date)
        if self.end_date:
            self.end_date = dateParser.parse_date(self.end_date)

        # Parse fixed leg
        if self.fixed_rate is not None:
            self.fixed_rate = float(self.fixed_rate)

        # Parse floating leg
        if self.floating_spread is not None:
            self.floating_spread = float(self.floating_spread)
        if self.fixing_days is not None:
            self.fixing_days = int(self.fixing_days)

        # Parse premium if provided
        if self.premium_amount is not None:
            self.premium_amount = float(self.premium_amount)
        if self.premium_currency is not None:
            self.premium_currency = self.premium_currency.upper()
        if self.premium_pay_date is not None:
            self.premium_pay_date = dateParser.parse_date(self.premium_pay_date)

        # Parse boolean
        if isinstance(self.is_in_arrears, str):
            self.is_in_arrears = self.is_in_arrears.lower() in ('true', 'yes', '1')
        else:
            self.is_in_arrears = bool(self.is_in_arrears)

    def _validate_input(self):
        """Validate trade parameters."""
        # Validate option parameters
        if self.exercise_style not in ['European', 'Bermudan', 'American']:
            raise ValueError(f"Invalid exercise style: {self.exercise_style}. Must be European, Bermudan, or American")

        if self.long_short not in ['Long', 'Short']:
            raise ValueError(f"Invalid long/short: {self.long_short}. Must be Long or Short")

        if self.settlement_type not in ['Cash', 'Physical']:
            raise ValueError(f"Invalid settlement type: {self.settlement_type}. Must be Cash or Physical")

        if self.payer_receiver not in ['Payer', 'Receiver']:
            raise ValueError(f"Invalid payer/receiver: {self.payer_receiver}. Must be Payer or Receiver")

        # Validate exercise dates based on style
        exercise_date_list = [d.strip() for d in self.exercise_dates.split(',')]

        if self.exercise_style == 'European':
            if len(exercise_date_list) != 1:
                raise ValueError(f"European swaption must have exactly 1 exercise date, got {len(exercise_date_list)}")
        elif self.exercise_style == 'American':
            if len(exercise_date_list) != 2:
                raise ValueError(f"American swaption must have exactly 2 exercise dates (start and end), got {len(exercise_date_list)}")
        elif self.exercise_style == 'Bermudan':
            if len(exercise_date_list) < 2:
                raise ValueError(f"Bermudan swaption must have at least 2 exercise dates, got {len(exercise_date_list)}")

        # Validate underlying swap parameters
        if self.notional is not None and self.notional <= 0:
            raise ValueError("Notional must be positive")

        if self.start_date and self.end_date:
            if self.start_date >= self.end_date:
                raise ValueError("End date must be after start date")

            # Parse and validate exercise dates are before swap start
            for exercise_date_str in exercise_date_list:
                exercise_date = dateParser.parse_date(exercise_date_str.strip())
                if exercise_date <= self.start_date:
                    raise ValueError(f"Exercise date {exercise_date_str} must be after swap start date")

        if self.fixing_days is not None and self.fixing_days < 0:
            raise ValueError("Fixing days must be non-negative")

        # Validate premium - if any premium field is provided, all must be provided
        premium_fields = [self.premium_amount, self.premium_currency, self.premium_pay_date]
        premium_count = sum(1 for f in premium_fields if not self._is_empty(f))
        if premium_count > 0 and premium_count < 3:
            raise ValueError("If premium is provided, all premium fields (amount, currency, pay_date) must be specified")

    @staticmethod
    def _is_empty(value):
        """
        Check if a value is empty (None, NaN, or empty string).

        Args:
            value: Value to check

        Returns:
            True if value is None, NaN, or empty string
        """
        if value is None:
            return True
        # Check for NaN (works for float NaN from pandas)
        try:
            if isinstance(value, float) and math.isnan(value):
                return True
        except (TypeError, ValueError):
            pass
        # Check for empty string
        if str(value).strip() == "":
            return True
        return False

    def _build_trade_dataframe(self):
        """
        Build pandas DataFrame with Swaption trade data for ORE.

        Returns:
            tuple: ('Swaption', pd.DataFrame with trade parameters)
        """
        trade_data = {
            'TradeId': [self.__name__],
            'TradeType': [self.__name__],
            'CounterParty': ['CPTY_A'],

            # Option parameters
            'LongShort': [self.long_short],
            'Style': [self.exercise_style],
            'Settlement': [self.settlement_type],
            'ExerciseDates': [self.exercise_dates],  # Comma-separated string

            # Underlying swap
            'Currency': [self.currency],
            'Notional': [self.notional],
            'StartDate': [dateParser.ql2pydate(self.start_date).strftime('%Y-%m-%d')],
            'EndDate': [dateParser.ql2pydate(self.end_date).strftime('%Y-%m-%d')],
            'Calendar': [self.calendar],
            'PaymentConvention': [self.payment_convention],
            'PayerReceiver': [self.payer_receiver],

            # Fixed leg
            'FixedRate': [self.fixed_rate],
            'FixedTenor': [self.fixed_tenor],
            'FixedDayCounter': [self.fixed_day_counter],

            # Floating leg
            'FloatingIndex': [self.floating_index],
            'FloatingTenor': [self.floating_tenor],
            'FloatingDayCounter': [self.floating_day_counter],
            'FloatingSpread': [self.floating_spread],
            'IsInArrears': [str(self.is_in_arrears).lower()],
            'FixingDays': [self.fixing_days]
        }

        # Add premium fields if provided
        if not self._is_empty(self.premium_amount):
            trade_data['PremiumAmount'] = [self.premium_amount]
            trade_data['PremiumCurrency'] = [self.premium_currency]
            trade_data['PremiumPayDate'] = [dateParser.ql2pydate(self.premium_pay_date).strftime('%Y-%m-%d')]

        trade_df = pd.DataFrame(trade_data)

        return (self.__name__, trade_df)

    def _filter_epe_dates(self, exposure_df):
        """
        Filter exposure DataFrame to include only relevant dates.

        Args:
            exposure_df (pd.DataFrame): DataFrame with exposure data including 'Date' column.
        Returns:
            pd.DataFrame: Filtered DataFrame.
        """
        exposure_df['Date'] = pd.to_datetime(exposure_df['Date'])
        start_date = dateParser.ql2npdate(self.start_date)
        end_date = dateParser.ql2npdate(self.end_date)
        filtered_df = exposure_df[(exposure_df['Date'] >= start_date) & (exposure_df['Date'] <= end_date)]
        return filtered_df
